controllers/secret_controller.go

```diff
@@ -0,0 +1,174 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// We need rights to update the target cluster's secrets, not necessarily this cluster
+// +kubebuilder:rbac:groups=core,namespace="k8ssandra",resources=secrets;replicatedsecrets,verbs=get;list;watch;update;create;delete
+
+type SecretSyncController struct {
+	ClientCache *clientcache.ClientCache
+	selectors   map[string]types.NamespacedName
+}
+
+func (s *SecretSyncController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	client, err := s.ClientCache.GetLocalClient()
+	if err != nil {
+		logger.Error(err, "Failed to fetch local client", "ReplicatedSecret", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	rsec := &api.ReplicatedSecret{}
+	if err := client.Get(ctx, req.NamespacedName, rsec); err != nil {
+		if errors.IsNotFound(err) {
+			// Does not exist in managed cluster - so we don't copy it
+			// TODO Unless we want to delete it from target clusters?
+			return reconcile.Result{}, nil
+		}
+		logger.Error(err, "Failed to get replicated secret", "ReplicatedSecret", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	// Add the new matcher rules also to our cache if not found
+	selector, err := metav1.LabelSelectorAsSelector(&rsec.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "Failed to transform to label selector", "ReplicatedSecret", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	if _, found := s.selectors[selector.String()]; !found {
+		s.selectors[selector.String()] = types.NamespacedName{Name: rsec.Name, Namespace: rsec.Namespace}
+	}
+
+	// Fetch all the secrets that match the ReplicatedSecret's rules
+	secrets, err := s.fetchAllMatchingSecrets(selector)
+	if err != nil {
+		logger.Error(err, "Failed to fetch linked secrets", "ReplicatedSecret", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	remoteClients, err := s.ClientCache.GetRemoteClients()
+	if err != nil {
+		logger.Error(err, "Failed to fetch remote clients", "ReplicatedSecret", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	for ctxName, remoteClient := range remoteClients {
+		// TODO Only replicate to clusters that are in the ReplicatedSecret's context
+
+		if remoteClient == client {
+			continue
+		}
+
+		// Iterate all the matching secrets
+		for _, sec := range secrets {
+			fetchedSecret := &corev1.Secret{}
+			if err := remoteClient.Get(ctx, types.NamespacedName{Name: sec.Name, Namespace: sec.Namespace}, fetchedSecret); err != nil {
+				if errors.IsNotFound(err) {
+					// Create it
+					copiedSecret := sec.DeepCopy()
+					copiedSecret.ResourceVersion = ""
+					if err := remoteClient.Create(ctx, copiedSecret); err != nil {
+						logger.Error(err, "Failed to sync secret to target cluster", "Secret", copiedSecret.Name, "Context", ctxName)
```

Comment:
(while outdated lines, the question is valid). Do you think it's safe if we keep replicating some secrets to some destination, but not others? How do we detect it's only a temporary issue and not something we should bail out from and wait for reconcile? Isn't continuing after problems a quite close to the "retry" idea?

Reply:
>Do you think it's safe if we keep replicating some secrets to some destination, but not others?

In general yes unless there is some sort of dependency between the objects being created and replicated. Suppose the spec tells us to replicate to 5 clusters and of those cluster is not reachable. Do I think we should replicate to the other 4? Absolutely.

>How do we detect it's only a temporary issue and not something we should bail out from and wait for reconcile?

We are dealing with secrets that are independent from one another. Contrast this with Cassandra pods where order does matter. When initializing the cluster we have to start seed nodes first. Furthermore, non-seed nodes must be bootstrapped serially.

>Isn't continuing after problems a quite close to the "retry" idea?

I don't think so since we are talking about creating different objects vs retrying the same one.